Recon Invariants

nican0r opened the pull request:
This PR adds a scaffolding and properties defined in the engagement with Recon. 

It also renames the `recon-core` -> `recon-vaults` for consistency with the CV, CP convention being used.

Todo:
- [x] scaffolding
- [x] full coverage
- [x] global properties
- [x] inline properties

Latest runs confirming full coverage (7/3):
- [CP](https://getrecon.xyz/shares/36cd15c9-2064-4a45-b0eb-cd0c1e3c7609)
- [CV](https://getrecon.xyz/shares/9fc4069d-e4d5-4ed7-9db0-a2e509314ad3)

lemunozm commented:
Hi @nican0r thanks for your work!

Could I ask you if you can fix the dependencies to ease reviews in the PR? Right now, the dependencies files are added to the main repo.

nican0r commented:
> Hi @nican0r thanks for your work!
> 
> Could I ask you if you can fix the dependencies to ease reviews in the PR? Right now, the dependencies files are added to the main repo.

Hi @lemunozm, sorry about that, should've checked building locally before pushing, I've fixed the dependency issues and modified the suite to work with the latest changes to `main`.

hieronx commented on remappings.txt:
```diff
@@ -1,3 +1,5 @@
 forge-gas-snapshot/=lib/forge-gas-snapshot/src/
+@recon/=lib/setup-helpers/src/
 @chimera/=lib/chimera/src/
+@recon/=lib/setup-helpers/src/
```
```suggestion
```

hieronx replied:
It's duplicated now

lemunozm commented on test/pools/fuzzing/recon-pools/README.md:
```diff
@@ -0,0 +1,18 @@
+The current setup uses one instance of the 
+`SingleShareClass` if the relationship between pools and Pools is one-to-many, more will need to be added.
+
+The `Gateway` contract is mocked since we're not actually using it to call the `PoolManager` functions and the extra transferring of gas adds unnecessary complexity to the tester.
```
Not sure if I understand correctly this last line, but the `Gateway` actually calls to `PoolManager` when we receive messages. It calls through [this interface](https://github.com/centrifuge/protocol-v3/blob/main/src/pools/interfaces/IPoolManager.sol#L173) that `PoolManager` implements.

nican0r replied:
Sorry could have been clearer, what I meant to say is that our fuzzing suite assumes the `Gateway` behaves correctly as there isn't much interesting logic there to test as it just mostly acts to forward messages received from the vaults and send messages to the vault (from my understanding). 

So I've setup the test suite so that calls that would normally be made by the `Gateway` to the `PoolManager` are made directly by the fuzzer instead so the fuzzer doesn't have to work around message handling in the `Gateway` which is assumed to be correct. The functions that are exposed to the fuzzer to do this are [here](https://github.com/centrifuge/protocol-v3/blob/442cff7f4a4048b228024740c671a020d4222c10/test/pools/fuzzing/recon-pools/targets/AdminTargets.sol#L132-L175).

I then mocked the `Gateway` [here](https://github.com/centrifuge/protocol-v3/blob/feat/recon-invariants/test/pools/fuzzing/recon-pools/mocks/MockGateway.sol) that's used in the setup and called by the other contracts to send messages back to the vaults side since the more recent changes to the `Gateway` regarding forwarding gas would require a decent amount of changes to the existing test suite that wouldn't actually add any more interesting state exploration because this is just forwarding messages out to the vaults side. 

Will add this more detailed description to the readme, pls lmk if there's anything still unclear though @lemunozm

lemunozm replied:
Makes perfect sense!! Thanks for the long explanation 🙌🏻

lemunozm reviewed (commented):
I did a review regarding things I know and I can verify, and by now, everything looks good to me!

wischli commented on test/pools/fuzzing/recon-pools/targets/AdminTargets.sol:
```diff
@@ -0,0 +1,211 @@
+
+    // === MultiShareClass === //
+
+    function singleShareClass_claimDepositUntilEpoch(PoolId poolId, ShareClassId shareClassId_, AssetId depositAssetId, uint32 endEpochId) public updateGhosts asAdmin {
```
Nit: Still uses `singleShareClass` namespace. We can fix this in a follow-up one though, definitely not a blocker. Just wanted to make you aware for future work.

nican0r replied:
Good catch, will fix this

wischli commented on recon.json:
```diff
@@ -0,0 +1,632 @@
+    ],
+    "separated": true
+  },
+  "out/SingleShareClass.sol/SingleShareClass.json": {
```
I suppose this file is auto-generated?

```suggestion
  "out/MultiShareClass.sol/MultiShareClass.json": {
```

wischli commented on test/pools/fuzzing/recon-pools/README.md:
```diff
@@ -0,0 +1,18 @@
+The `adapter` set on the `Gateway` is set to the address of a `MockAdapter` in the `Setup` contract. This is a simplification to allow the necessary logic but it shouldn't forward calls made to the `execute` function because these should be made directly via the `PoolRouterTargets` contract.
+
+The current setup uses one instance of the 
+`SingleShareClass` if the relationship between pools and Pools is one-to-many, more will need to be added.
```
Nit
```suggestion
`MultiShareClass` if the relationship between pools and Pools is one-to-many, more will need to be added.
```

wischli commented on test/pools/fuzzing/recon-pools/TargetFunctions.sol:
```diff
@@ -0,0 +1,418 @@
+        return (poolId, scId);
+    }
+
+    function shortcut_deposit_and_cancel(
```
Note: Another possible flow would be if the user just requested a deposit and immediately follows up with the cancel. However, I believe it's sufficient to regard your scenario as the only one as its more realistic/common.

nican0r replied:
Good point, when I tried adding this in though I'm getting a revert with `ClaimDepositRequired` due to this line: 
<img width="978" alt="image" src="https://github.com/user-attachments/assets/543cce02-f6a3-4ba5-9be7-32873453fe8a" />

So it seems like user needs to claim before cancelling their deposit request, is that correct?

wischli replied:
> So it seems like user needs to claim before cancelling their deposit request, is that correct?

Only if in between executing the `requestDeposit` and `cancelDepositRequest`, the FM at least approved deposits via `approveDeposits`. Then, `userOrder.lastUpdate <= latestApproval`, such that the user needs to claim (after the FM also issued new shares). The reason is that after an approval happens, some share of the users pending amount was approved which they must not be able to mutate until they claimed such that they have 0 pending approved.

The unit test `testCancelDepositRequest` proves that a user can immediately cancel after requesting a deposit.

nican0r replied:
Ah, I see now. I think this is a case worth exploring so added a shortcut for it [here](https://github.com/centrifuge/protocol-v3/blob/3fb3837b2251c6d2062ba89a99d09d2ca411f286/test/pools/fuzzing/recon-pools/TargetFunctions.sol#L140-L162)

wischli commented on test/pools/fuzzing/recon-pools/TargetFunctions.sol:
```diff
@@ -0,0 +1,418 @@
+        
+        // claim redemption as actor
+        bytes32 investor = Helpers.addressToBytes32(_getActor());
+        poolManager_claimRedeem(poolId, scId, isoCode, investor);
```
Nit:
```suggestion
        // claim redemption as actor
        shortcut_claim_redemption(poolId, scId, isoCode);
```

wischli replied:
Nit:
```suggestion
        // claim redemption as actor
        shortcut_claim_redemption(poolId, scId, isoCode);
```

wischli commented on test/pools/fuzzing/recon-pools/TargetFunctions.sol:
```diff
@@ -0,0 +1,418 @@
+        
+
+        // claim redemption as actor
+        shortcut_claim_redemption(poolId, scId, isoCode);
```
Q: Why not use the shortcut? Because of `stack too deep`?
```suggestion
        shortcut_redeem_and_claim(poolId, scId, shareAmount, isoCode, _getMultiShareClassMetrics(scId), navPerShare, isIdentityValuation);
```

nican0r replied:
Yes exactly, replacing with the shortcut gives a stack too deep error so I used the current implementation as a workaround

wischli commented on test/pools/fuzzing/recon-pools/Properties.sol:
```diff
@@ -0,0 +1,54 @@
+                    gte(
+                        multiShareClass.pendingRedeem(scId, assetId), 
+                        pendingUserRedemption, 
+                        "cancelled redemption is greater than requested redemption"
```
Nit: Comment is off
```suggestion
                        "user redemption is greater than total redemption"
```

wischli reviewed (commented):
Thanks for your great work to get this running! A few nit comments from my side, not a blocker for sure.

nican0r committed:
feat: `cancelRequests` absolute values should never be higher than `user.order` property

nican0r committed:
chore: update comments for properties

nican0r committed:
feat: properties to catch reverts in _updateDepositRequest and _updateRedeemRequest

nican0r committed:
fix: shortcut_deposit_cancel_redemption

nican0r committed:
feat: shortcut to deposit and cancel all at once

nican0r committed:
chore: fix property checks for handlers using try/catch

nican0r committed:
fix: clearing queued calls before calling shortcuts to make reading reproducers simpler

nican0r committed:
chore: add reproducers

nican0r committed:
chore: remove canary

nican0r committed:
chore: clamping decimal values in shortcut functions

nican0r committed:
fix: Property: The total pending deposit amount pendingDeposit[..] is always >= the sum of pending user deposit amounts depositRequest[..] to be inlined

nican0r committed:
fix: indexing of share classes in properties to skip the first which is never assigned

nican0r committed:
chore: remove unnecessary inlined checks

nican0r committed:
fix: extra precondition for arithmetic checks

nican0r committed:
chore: removing transient properties that always break

nican0r committed:
chore: remove duplicate property

nican0r committed:
chore: remove duplicate property

nican0r committed:
fix: property_total_pending_redeem_geq_sum_pending_user_redeem

nican0r committed:
chore: better handling of _setEpochIncrement in TFs

nican0r committed:
feat: property_holdings_balance_equals_escrow_balance

nican0r committed:
feat: property_assetRegistry_balance_leq_escrow_balance

nican0r committed:
chore: comment for testing

nican0r committed:
chore: formatting for multiple testers

nican0r committed:
chore: external calls handled by asAdmin modifier

nican0r committed:
chore: update gitignore

nican0r committed:
Fix/cv tester latest (#190)

* feat: updated CV tester

* fix: adapting latest suite to use actors + managers

* fix: use of asset deployed through manager throughout the suite

nican0r committed:
feat: missing setup for remaining coverage

nican0r committed:
fix: add missing ManagerTargets

nican0r committed:
fix: expanding coverage

nican0r committed:
fix: vault_redeem property to work with multi-asset setup

nican0r committed:
fix: invariant_global_2

nican0r committed:
fix: global properties to use vault asset

nican0r committed:
fix: minting asset to escrow

nican0r committed:
fix: deposited asset and tracking + reproducer

nican0r committed:
Merge remote-tracking branch 'origin/main' into feat/recon-invariants

nican0r committed:
fix: resolve build errors after merging in main

nican0r committed:
chore: naming consistency + folder structure cleanup

nican0r committed:
fix: coverage

hieronx commented on test/pools/fuzzing/recon-pools/Properties.sol:
```diff
@@ -0,0 +1,402 @@
+    /// === Global Properties === ///
+
+    function property_unlockedPoolId_transient_reset() public {
+        eq(_after.ghostUnlockedPoolId.raw(), 0, "unlockedPoolId not reset");
```
These are neat!

nican0r committed:
fix: update handlers and setup to latest interface for PoolRouter

nican0r committed:
fix: handling new interface for createHolding function

nican0r committed:
feat: targets for new functions

nican0r committed:
fix: more coverage blockers

nican0r committed:
Merge remote-tracking branch 'origin/main' into feat/recon-invariants

nican0r committed:
fix: build errors from latest merge with main

nican0r committed:
feat: handler for updateRestriction

nican0r committed:
chore: refactor inputs to use underlying type + set name and symbol as static for shareClass

nican0r committed:
chore: removing data params and setting to constant

nican0r committed:
Merge remote-tracking branch 'origin/main' into feat/recon-invariants

nican0r committed:
feat: missing multiShareClass handlers

nican0r committed:
chore: renaming poolRouter to hub in test suite

nican0r committed:
chore: update test runner command in readme

nican0r committed:
feat: handlers for decreaseShareIssuance and increaseShareIssuance

nican0r committed:
Merge branch 'main' into feat/recon-invariants

nican0r committed:
fix: build errors

nican0r committed:
chore: resolving build warnings

nican0r committed:
chore: rename recon-vault to recon-core

nican0r committed:
Property: accounting.accountValue should never revert

nican0r committed:
Property:  account.totalDebit and account.totalCredit is always less than uint128(type(int128).max)

nican0r committed:
feat: differential fuzz test

nican0r committed:
chore: resolve warnings

nican0r committed:
chore: extra precondition to resolve incorrect spec in broken property

nican0r committed:
Merge branch 'main' into feat/recon-invariants

nican0r committed:
chore: fix warnings

nican0r committed:
Property: Any decrease in valuation should not result in an increase in accountValue

nican0r committed:
Merge remote-tracking branch 'origin/main' into feat/recon-invariants

nican0r committed:
fix: naming of HubTargets

hieronx committed:
Merge branch 'main' of github.com:centrifuge/protocol-v3 into feat/recon-invariants

hieronx committed:
Fix warnings

nican0r committed:
Merge remote-tracking branch 'origin/main' into feat/recon-invariants

nican0r committed:
refactor: separate out toggle target functions

nican0r committed:
fix: dependencies as submodules

nican0r committed:
Merge remote-tracking branch 'origin/main' into feat/recon-invariants

nican0r committed:
fix: missing coverage in updateHoldingAmount when decreasing

nican0r committed:
fix: calling unclamped handler from clamped

nican0r committed:
feat: clamped handlers for missing coverage on claimDepositUntilEpoch + claimRedeemUntilEpoch

nican0r committed:
Merge remote-tracking branch 'origin/main' into feat/recon-invariants

nican0r committed:
fix: updating handlers to latest Hub changes

nican0r committed:
chore: missing coverage over updateHolding

nican0r committed:
feat: high level accounting properties

nican0r committed:
chore: more clamped handlers

nican0r committed:
chore: cast bytes32(salt) from uint256 for easier debugging of repros

nican0r committed:
fix: preconditions for properties in hub_depositRequest

nican0r committed:
chore: replacing D18 in handlers because of errors with echidna repros

nican0r committed:
chore: remove poorly defined property

nican0r committed:
fix: precondition for hub_claimDeposit function

nican0r committed:
Merge branch 'main' into feat/recon-invariants

nican0r committed:
fix: build errors for latest interface

nican0r committed:
chore: change pools test to correct location

nican0r committed:
chore: fixing coverage for latest vault changes

nican0r committed:
chore: clamping for vaults

nican0r committed:
chore: precondition in hub_depositRequest

nican0r committed:
fix: missing coverage

nican0r committed:
chore: rearrange reproducers

nican0r committed:
chore: fix setup for new dependencies

nican0r committed:
chore: vault solvency property + optimization test

nican0r committed:
feat: property_totalAssets_insolvency_only_increases

nican0r committed:
feat: doomsday ppfs properties

nican0r committed:
feat: property_pricePerShare_never_changes_after_user_operation

nican0r committed:
feat: doomsday_impliedPricePerShare_never_changes_after_user_operation

nican0r committed:
chore: additional checks for before/after vars

nican0r committed:
fix: missing withdrawal coverage in vaults

nican0r committed:
fix: property_totalAssets_solvency precondition

nican0r committed:
fix: remove cross-chain transfer functions that cause false positives

nican0r committed:
fix: ppfs properties

nican0r committed:
feat: more doomsday deposit/withdraw properties

nican0r committed:
fix: additional preconditions for doomsday properties

nican0r committed:
fix: build error

nican0r committed:
chore: fix properties

nican0r committed:
chore: fix clamped handlers to call unclamped

nican0r committed:
chore: add reproducers and rearrange tests

nican0r committed:
fix: asset tracking

nican0r committed:
Merge remote-tracking branch 'origin/main' into feat/recon-invariants

nican0r committed:
fix: build errors after pulling latest main

nican0r committed:
fix: stateless properties

nican0r committed:
fix: property_totalAssets_solvency

nican0r committed:
fix: property_E_1

nican0r committed:
fix: using CastLib for consistency

nican0r committed:
feat: more vault properties

nican0r committed:
End-To-End Tester (#276)

* feat: inheritance structure for reusable properties

* Revert "feat: inheritance structure for reusable properties"

This reverts commit 3c2d2d99e6724deed35b42c1149ffa5b5ce275b8.

* feat: initial setup for end-to-end tester

* fix: coverage on shortcut_deployNewTokenPoolAndShare

* fix: missing coverage

* fix: MockMessageDispatcher to correctly handle message sending

* fix: vaultcallbacks to be made through appropriate channels

* chore: add readme

* feat: SCM properties

* fix: create accounts and holding in deployment shortcut

* feat: increasing coverage on e2e tester

* fix: additional missing coverage

* fix: remaining coverage

* fix: stateless tests

* fix: property_totalAssets_solvency

* fix: property_E_1

* fix: build error

* fix: adding shortcuts and approvals for coverage on claimRedeem

* feat: maxDeposit/Redeem properties

* refactor: previous max properties for checking amounts up to max

* feat: difference properties

* feat: maxDeposit clears all orders

* feat: more vault properties

* feat: more shortcuts for better coverage

* chore: cleaning up shortcuts

* feat: escrow solvency property

* fix: price setting handlers

* feat: property_price_per_share_overall

nican0r committed:
Merge branch 'main' into feat/recon-invariants

nican0r committed:
Merge branch 'feat/recon-invariants' of https://github.com/centrifuge/protocol-v3 into feat/recon-invariants

nican0r committed:
chore: partially resolving build errors for latest changes

nican0r committed:
fix: build errors

nican0r committed:
fix: property_escrow_solvency

nican0r committed:
chore: cleanup comment

nican0r committed:
fix: global properties for latest interface

nican0r committed:
fix: inline properties for latest interface

nican0r committed:
fix: vars in BeforeAfter for latest interface

nican0r committed:
fix: lib error when deploying contracts

nican0r committed:
fix: error in setup

nican0r committed:
fix: coverage issues

nican0r committed:
fix: coverage issues in vaults tester

nican0r committed:
fix: MockMessageDispatcher for consistency

nican0r committed:
fix: missing filings for messageDispatcher in Hub setup

nican0r committed:
fix: messageDispatcher dependencies in hub setup

nican0r committed:
fix: add handlers for missing targets

nican0r committed:
chore: clamping for covering redemptions

nican0r committed:
fix: clamping to get coverage of redemptions + cleanup console logs

nican0r committed:
feat: add sync vault to e2e tester

nican0r committed:
chore: removing unnecessary SCM wrapper

nican0r committed:
fix: inlined escrow properties for sync vaults

nican0r committed:
fix: inline property checks

nican0r committed:
chore: more shortcuts

nican0r committed:
chore: updating docs

nican0r committed:
chore: more shortcuts for better exploration of cancelled states

nican0r committed:
chore: removing Hub handlers that are no longer needed

nican0r committed:
chore: remove console log from AsyncRequestManager

nican0r committed:
chore: shortcuts for claiming cancelled amounts

nican0r committed:
chore: removing false positives

nican0r committed:
chore: shortcuts for minting/depositing in sync vaults

nican0r committed:
chore: add missing syncRequestManager_setValuation handler

nican0r committed:
chore: update readme

nican0r committed:
chore: fixing false positives

nican0r committed:
fix: false positive broken prop

nican0r committed:
chore: fixing broken properties

nican0r committed:
fix: false positive broken properties

nican0r committed:
chore: reproducers

nican0r committed:
chore: logging statements in test

nican0r committed:
fix: false positives

nican0r committed:
feat: ReconPoolManager for handling multiple deployed pools

nican0r committed:
fix: clamping created pools using manager

nican0r committed:
feat: manager for scId

nican0r committed:
feat: adding ReconAssetIdManager

nican0r committed:
chore: manager for vaults

nican0r committed:
feat: manager for share tokens

nican0r committed:
chore: consolidating ghosts into BeforeAfter

nican0r committed:
chore: update property names and descriptions

nican0r committed:
refactor: properties

nican0r committed:
fix: coverage blockers in BeforeAfter updates

nican0r committed:
fix: property_eligible_user_deposit_amount_leq_deposit_issued_amount

nican0r committed:
fix: updating ghosts in handlers

nican0r committed:
fix: false positives

nican0r committed:
fix: transfer false positives

nican0r committed:
fix: properties to use current vault as source of truth

nican0r committed:
chore: fixing handlers

nican0r committed:
chore: add missing doomsday targets

nican0r committed:
fix: false positive properties

nican0r committed:
fix: more false positives

nican0r committed:
chore: expand repro

nican0r committed:
chore: property fix and reproducer

nican0r committed:
chore: properties table

nican0r committed:
Update properties-table.md

fix incorrect property function names

nican0r committed:
chore: simplifying property + repros

nican0r committed:
Merge remote-tracking branch 'origin/spearbit-fixreview' into feat/recon-invariants

nican0r committed:
fix: build errors after merging in PR355

nican0r committed:
fix: coverage blockages due to latest changes

nican0r committed:
fix: remove unnecessary loops in properties

nican0r committed:
fix: false positives

nican0r committed:
fix: SpokeTargets name

nican0r committed:
chore: update readme

nican0r committed:
chore: add repros

nican0r committed:
chore: additional logging to test_token_transfer_9 repro

nican0r committed:
fix: false positives

nican0r committed:
fix: false positives

nican0r committed:
feat: property_total_issuance_soundness

nican0r committed:
fix: property_escrow_share_balance

nican0r committed:
chore: adding reproducers

nican0r committed:
fix: false positives + optimization tests

nican0r committed:
refactor: fixing shortcuts that immediately approve issue/revoke before cancel

nican0r committed:
fix: property_total_issuance_soundness

nican0r committed:
chore: shortcut for queueing deposits

wischli committed:
fix: total redeem property (#404)

* fix: total redeem property

* fix: property_sum_pending_user_deposit_geq_total_pending_deposit as global property

---------

Co-authored-by: nelson-pereira8 <94120714+nican0r@users.noreply.github.com>

nican0r committed:
fix: build errorz

nican0r committed:
feat: add BalanceSheetTargets

nican0r committed:
fix: ghost share tracking for shares minted by BalanceSheet

nican0r committed:
fix: accidentally removed variable from property

nican0r committed:
chore: add updateGhosts modifier to BalanceSheetTargets

nican0r committed:
fix: false positives

nican0r committed:
chore: more property descriptions

nican0r committed:
chore: update properties table

nican0r committed:
chore: handler formatting

nican0r committed:
fix: using only values derived from currently set vault in handlers

nican0r committed:
chore: separate foundry test file for sanity tests

nican0r committed:
chore: newest reproducers

nican0r committed:
chore: adding all targets to readme

nican0r committed:
chore: notes on reproducers

nican0r committed:
fix: property implementations

nican0r committed:
fix: false positives

nican0r committed:
Merge remote-tracking branch 'origin/spearbit-fixreview' into feat/recon-invariants

nican0r committed:
chore: fix build errors after merging spearbit-fixreview branch

nican0r committed:
chore: missing Hub handlers

nican0r committed:
chore: consistency of handler names

nican0r committed:
chore: more naming consistency

nican0r committed:
fix: get rid of loops in BeforeAfter

nican0r committed:
fix: breaking setup in tests

nican0r committed:
chore: update properties table to reflect temporarily removed properties

nican0r committed:
fix: missing coverage due to latest updates

nican0r committed:
fix: setup for wards on tokenFactory

nican0r committed:
fix: broken property

nican0r committed:
chore: reproducers

nican0r committed:
chore: ghost tracking to fix property_actor_pending_and_queued_redemptions false positive

nican0r committed:
fix: false positives

nican0r committed:
fix: false positives

nican0r committed:
fix: false positives

nican0r committed:
chore: reproducer for maxRedeem issue

nican0r committed:
chore: reproducer

nican0r committed:
fix: comment for test_asyncVault_maxRedeem_8

nican0r committed:
chore: ppfs properties

nican0r committed:
fix: ppfs property implementation

nican0r committed:
chore: reproducer

nican0r committed:
feat: optimize_maxDeposit_difference

nican0r committed:
chore: BeforeAfter tracking for PPFS properties

nican0r committed:
chore: additional precondition for property_totalAssets_solvency

nican0r committed:
fix: property_total_issuance_soundness

nican0r committed:
chore: remove irrelevant property

nican0r committed:
fix: property_eligible_user_redemption_amount_leq_approved_asset_redemption_amount secondary check

nican0r committed:
chore: updating properties table

nican0r committed:
chore: repros and cleanup

nican0r committed:
fix: false positive

nican0r committed:
fix: optimization property for maxDeposit

hieronx commented:
Closing in favour of https://github.com/centrifuge/protocol-v3/pull/468
